Fix `#recognize` inconsistency with `#serve`

skipkayhil opened the pull request:
### Motivation / Background

When routing a request, `#find_routes` will iterate through potential routes for the request and yield them to the passed block. `#serve` and `#recognize` both used `#find_routes`, but ended up handling things slightly differently. Both methods would modify the `script_name` and `path_info` of the request if the current route is not anchored. However, only `#serve` used `#chomp` and would restore the request's original `script_name` and `path_info` after each iteration. This inconsistency can cause issues as `#recognize` can end up not matching them even though `#serve` would.

### Detail

This commit fixes the issue by unifying more of the implementations of `#serve` and `#recognize`. Now, `#find_routes` implements the mutation/restoration of request params for both `#serve` and `#recognize` so that their behavior is consistent and request mutation isn't leaked between iterations.

One test had to be changed because it depended on the request mutation leaking outside of `#recognize`, so it was updated to make assertions during `#recognize` instead.

### Additional information

Fixes #31152

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil committed:
Fix `#recognize` inconsistency with `#serve`

When routing a request, `#find_routes` will iterate through potential
routes for the request and yield them to the passed block. `#serve` and
`#recognize` both used `#find_routes`, but ended up handling things
slightly differently. Both methods would modify the `script_name` and
`path_info` of the request if the current route is not anchored.
However, only `#serve` used `#chomp` and would restore the request's
original `script_name` and `path_info` after each iteration. This
inconsistency can cause issues as `#recognize` can end up not matching
them even though `#serve` would.

This commit fixes the issue by unifying more of the implementations of
`#serve` and `#recognize`. Now, `#find_routes` implements the
mutation/restoration of request params for both `#serve` and
`#recognize` so that their behavior is consistent and request mutation
isn't leaked between iterations.

One test had to be changed because it depended on the request mutation
leaking outside of `#recognize`, so it was updated to make assertions
during `#recognize` instead.

skipkayhil commented:
@byroot would you mind reviewing this one as well 🙇

byroot commented on actionpack/lib/action_dispatch/journey/router.rb:
```diff
@@ -27,45 +27,20 @@ def eager_load!
-
-          yield(route, parameters)
-        end
+      def recognize(rails_req, &block)
+        find_routes(rails_req, &block)
       end
```
Why not just use an alias? Is the method even still needed?

skipkayhil replied:
I wanted to keep the diff clean, but I agree that an alias (or inlining everything into `recognize`) makes sense to me.

I'll update with another commit

skipkayhil committed:
Inline find_routes into recognize

(Separate from the previous commit so that the original refactor is clear
in the diff)

Now that recognize just delegates to find_routes, there is no longer a
need for find_routes to exist separately.
